backward: Add dependency to just grab header

WillAyd opened the pull request:
backward-cpp distributes a a header-only library in addition to the module. The header-only library can be useful as a standalone dependency, especially on systems where the module may be difficult to compile (i.e. Windows)

https://github.com/bombela/backward-cpp?tab=readme-ov-file#installation

WillAyd commented:
That is a surprising amount of failures, although I don't believe any of them are related to this change (?)

dcbaker commented:
A couple of things I see. I think you'll need to include the `-DBACKTRACE_HAS_N` declare_dependency for the interface. which likely means taking the stuff out of the `add_project_rguments(...)`, sticking it in an array, and then adding it to the dependencies as necessary.

I'm not sure if backtrace has been updated since alphine was added. This is annoying, because alpine is MUSL based, and doesn't have an execinfo.h or libexecinfo, so we really need to have a check in the Meson that does something like (untested):
```meson
cpp = meson.get_compiler('cpp')
has_backtrace = cpp.has_header('execinfo.h') and cpp.has_header_symbol('execinfo.h', 'backtrace')
add_project_arguments('-DBACKWARD_HAS_BACKTRACE=@0@'.format(has_backtrace.to_int(), language : 'cpp')
```
to deal with this.

neheb commented:
Last I looked into this fixing the execinfo.h issue requires upstream changes/

dcbaker commented:
I just looked at the header to see that you could set `BACKTRACE_HAS_BACKTRACE=0`, but I'm not surprised to hear that path isn't well tested and doesn't work.

WillAyd commented:
Just rebased on top of https://github.com/mesonbuild/wrapdb/pull/2129

In the process I simplified this and took out any of the macro changes, since 2129 mostly dealt with those. I think there's still a gap with the `BACKWARD_HAS_BACKTRACE` macro and detection, but hoping we can leave that to another PR

bgilbert reviewed (commented):
I think this still needs to propagate the defines into the `declare_dependency()` `compile_args` as mentioned in the first paragraph of https://github.com/mesonbuild/wrapdb/pull/2072#issuecomment-2839710221.

WillAyd commented on subprojects/backward-cpp.wrap:
```diff
@@ -8,3 +8,4 @@ patch_directory = backward-cpp
 
 [provide]
 backward-cpp = backward_dep
+backward-cpp-interface = backward_cpp_interface_dep
```
done!
